elastic/elastic-agent
# Minimal ESS API client for integration/E2E testing framework

## What does this PR do?



This PR introduces a minimal Golang client for the [ESS API](https://www.elastic.co/guide/en/cloud/current/ec-restful-api.html).  Specifically, it implements just enough of the ESS API as needed by the converged integration/E2E framework:
- [Create a deployment](https://www.elastic.co/guide/en/cloud/current/Deployment_-_CRUD.html#create-deployment). 
- [Shut down a deployment](https://www.elastic.co/guide/en/cloud/current/Deployment_-_CRUD.html#shutdown-deployment).

Additionally, the client also includes two convenience methods for:
- Retrieving the status of the deployment, including the statuses of each of the deployment's components
- Checking if the deployment is ready for use, i.e. all its components have started

Note that the client introduced in this PR makes the following assumptions:
* It defaults to using the QA ESS environment; this is because deployments in this environment are automatically cleaned up every ~24 hours or so, which is ideal for testing — it won't leave behind orphaned deployments.
* It assumes that the deployments will be spun up in GCP.  This can be changed in the future, as needed.
* It spins up a deployment with just enough components and their instances for testing with ES and Fleet (specifically: one 1-node ES cluster, one Kibana instance, and one Integrations Server instance).

## Why is it important?

For the converged integration/E2E framework to be able to provision Elastic Stack deployments for testing.



## Checklist




## How to test this PR locally

1. _First time only_: Sign in to the [QA ESS user console](https://console.qa.cld.elstc.co/) and [obtain an API key](https://www.elastic.co/guide/en/cloud/current/ec-api-authentication.html).
2. Run the test included in this PR, `TestClient_CreateAndShutdownDeployment`, passing in the API key as an environment variable.
   ```
   ESS_API_KEY=XXXXXXXXXXXXXX go test ./pkg/testing/ess -run TestClient_CreateAndShutdownDeployment -v -count 1
   ```

   _Optional_: if you want to verify manually that the deployment is actually being created by the test, e.g. by looking in the QA ESS user console, ask the test to delay shutting down the deployment by passing the shutdown delay, in seconds, as another environment variable.
   ```
   ESS_API_KEY=XXXXXXXXXXXXXX ESS_CLIENT_TEST_SHUTDOWN_DELAY_SECONDS=30 go test ./pkg/testing/ess -run TestClient_CreateAndShutdownDeployment -v -count 1
   ```
